### ISSUE TITLE
cmd_jumpToNextSection cmd_jumpToPreviousSection not working

### ISSUE THREAD

**Issue by agroszer** [Komodo/KomodoEdit#3013]
Komodo IDE, version 11.0.0-alpha1, build 90546, platform linux-x86_64.
Built on Fri Aug 25 00:23:09 2017.

Did not reset codeintel DB before trying, so it might have some crap...

opened up https://github.com/zopefoundation/z3c.form/blob/master/src/z3c/form/button.py

saved, codeintel scan completed

`cmd_jumpToNextSection` or `cmd_jumpToPreviousSection` does not move the cursor to the next/prev class/def/whatever

it occasionally moves, but seemingly weird places

**Comment by cgchoffman** [comment]
Can you be more specific about when it doesn't work?  It seems to be working fine for me.

**Comment by cgchoffman** [comment]
er no, WHERE not when.  Where is your cursor in your code when it doesn't work?  Can you narrow it down to smaller example then an entire 300 line file?

**Comment by agroszer** [comment]
seems to work with latest releases and several codeintel DB rebuilds

it seems to be a tad slower than in KO10

also note, you renamed `cmd_jumpToPreviousSection` to `cmd_jumpToPrevSection`

**Comment by Naatan** [comment]
> and several codeintel DB rebuilds

This isn't a thing anymore in 11. The Database should be fully reliable. If you made it work it is very unlikely to have anything to do with a DB rebuild.

**Comment by Naatan** [comment]
> also note, you renamed cmd_jumpToPreviousSection to cmd_jumpToPrevSection

@cgchoffman can you verify if this will cause issues with any of our keybindings?

**Comment by cgchoffman** [comment]
> also note, you renamed cmd_jumpToPreviousSection to cmd_jumpToPrevSection

This effects all keybindings that have that command which appears to be all of them.

**Comment by cgchoffman** [comment]
> which appears to be all of them.

Correction, all except the `default` scheme, also I forgot to check the upgrade code.  This might be fixed there.

**Comment by cgchoffman** [comment]
> I forgot to check the upgrade code. This might be fixed there.

It is not.  cmd_jumpToPrevSection doens'nt have a keybinding in the default keybinding scheme.

**Comment by Naatan** [comment]
The core issue here seems to be the following type of statement in the file:

```
@zope.interface.implementer(interfaces.IButtonHandler)
```

For some reason this breaks Komodo's scope locating mechanic.

**Comment by mitchell-as** [comment]
Punting to 11.1. There are two issues here. One is that the Python parser does not process class decorators at all (it's fine with function decorators). Two is that the `getNextScope()` and `getPrevScope()` CodeIntel service functions do not function properly in the main scope -- they only function properly inside a sub-scope, like a class or function. Since our refactoring code depends on `getNextScope()` and `getPrevScope()`, attempting a fix here may cause regressions that are not acceptable for 11.0.1.

**Comment by mitchell-as** [comment]
@cgchoffman Tentatively co-assigning with you since you wrote the original `get*Scope()` code.

**Comment by cgchoffman** [comment]
Let me know what you need from me.  I didn't do anything special in the scope functions.  I just got the enclosing scope of the cursors location.  I don't know why that wouldn't work in the main scope.  If it functions differently than other scopes then it will have to be a special case for all cases that the cursor is in the main scope.

**Comment by mitchell-as** [comment]
The problematic lines are:

```
         caretScope = scope.resolveScope(line)
         if not caretScope or not isinstance(caretScope, AbstractSymbol):
             return
```
The main scope is not an instance of an AbstractSymbol. It's just an AbstractScope. We need to find the first AbstractSymbol in the main scope that comes after `line`.

**Comment by cgchoffman** [comment]
I'll be honest, I'm pretty sure I just copied that from other functions in that file because I don't know anything about how code intel was written and I had to get refactoring working.  This is likely why I was seeing issues with some cases of jumping to a scope.  If that shouldn't be there it would be your call to remove it then we just see if refactoring still works and go from there. 

I agree this should be 11.1 but my working knowledge of this code is not better than yours @mitchell-as .
